src/scheduler_utils.cc

```diff
@@ -247,41 +247,54 @@ PriorityQueue::PolicyQueue::TimeoutAt(size_t idx)
   }
 }
 
+bool
+PriorityQueue::PolicyQueue::ReadyForErasure()
+{
+  size_t total_size = Size() + rejected_queue_.size();
+  return !keep_instantiated_ && total_size == 0;
+}
+
 PriorityQueue::PriorityQueue()
-    : size_(0), front_priority_level_(0), last_priority_level_(0)
+    : size_(0), front_priority_level_(0), last_priority_level_(0),
+      default_policy_()
 {
-  inference::ModelQueuePolicy default_policy;
-  queues_.emplace(0, PolicyQueue(default_policy));
+  queues_.emplace(0, PolicyQueue(default_policy_, true));
   front_priority_level_ = queues_.begin()->first;
   ResetCursor();
 }
 
 PriorityQueue::PriorityQueue(
     const inference::ModelQueuePolicy& default_queue_policy,
     uint32_t priority_levels, const ModelQueuePolicyMap queue_policy_map)
-    : size_(0), last_priority_level_(priority_levels)
+    : size_(0), last_priority_level_(priority_levels),
+      default_policy_(default_queue_policy)
 {
+  // Permanently instantiate PolicyQueue with keep_instantiate=true
+  // to prevent them from being erased & created during scheduling
   if (priority_levels == 0) {
-    queues_.emplace(0, PolicyQueue(default_queue_policy));
+    // Only default policy is instantiated
+    queues_.emplace(0, PolicyQueue(default_policy_, true));
   } else {
-    for (uint32_t level = 1; level <= priority_levels; level++) {
-      auto it = queue_policy_map.find(level);
-      if (it == queue_policy_map.end()) {
-        queues_.emplace(level, PolicyQueue(default_queue_policy));
-      } else {
-        queues_.emplace(level, PolicyQueue(it->second));
-      }
+    // All priorities with user-given policy are instantiated. We do not
+    // permanently add default PolicyQueue because those will be dynamically
+    // created and erased to keep memory footprint low
+    for (auto it = queue_policy_map.begin(); it != queue_policy_map.end();
```

Comment:
For consistency with the other code in the core repo, can we use the shorter form a for loop here? E.g. for (const auto& queue_policy : queue_policy_map)

Reply:
Done, replaced it with a range-based for loop